lib/Sema/CSApply.cpp

```diff
@@ -3656,6 +3656,27 @@ namespace {
             : CheckedCastContextKind::ConditionalCast;
 
       auto fromType = cs.getType(sub);
+
+      // Special handle for literals conditional checked cast when they can be
+      // statically coerced to the cast type.
+      if (isa<LiteralExpr>(sub)) {
```

Comment:
Could you please leave behind a FIXME about revising this behavior?  This is a warning for a deeply unintuitive behavior that we need to revisit.

Reply:
Filed SR-12093 about this.

Reply:
> Could you please leave behind a FIXME about revising this behavior? This is a warning for a deeply unintuitive behavior that we need to revisit.

Sure :)